src/hotspot/share/opto/multnode.hpp

```diff
@@ -49,6 +49,89 @@ class MultiNode : public Node {
   ProjNode* proj_out(uint which_proj) const; // Get a named projection
   ProjNode* proj_out_or_null(uint which_proj) const;
   ProjNode* proj_out_or_null(uint which_proj, bool is_io_use) const;
+  uint number_of_projs(uint which_proj) const;
+  uint number_of_projs(uint which_proj, bool is_io_use) const;
+
+  // Run callback on all Proj projection from this node
+  template<class Callback> ProjNode* apply_to_projs(Callback callback) const {
+    DUIterator_Fast imax, i = fast_outs(imax);
+    return apply_to_projs(imax, i, callback);
+  }
+
+  // Same but with provided iterators
+  template<class Callback> ProjNode* apply_to_projs(DUIterator_Fast& imax, DUIterator_Fast& i, Callback callback) const {
+    return apply_to_projs_any_iterator<Callback, UsesIteratorFast>(UsesIteratorFast(imax, i, this), callback);
+  }
+
+  // Same but only for Proj node whose _con matches which_proj
+  template <class Callback> ProjNode* apply_to_projs(DUIterator_Fast& imax, DUIterator_Fast& i, Callback callback, uint which_proj) const;
+
+  // Same but with default iterators
+  template<class Callback> ProjNode* apply_to_projs(Callback callback, uint which_proj) const {
+    DUIterator_Fast imax, i = fast_outs(imax);
+    return apply_to_projs(imax, i, callback, which_proj);
+  }
+
+  // Same but for matching _con and _is_io_use
+  template <class Callback> ProjNode* apply_to_projs(Callback callback, uint which_proj, bool is_io_use) const;
+
+protected:
+
+  // Provide single interface for DUIterator_Fast/DUIterator for template method below
+  class UsesIteratorFast {
+    DUIterator_Fast& _imax;
+    DUIterator_Fast& _i;
+    const Node* _node;
+
+  public:
+    bool cont() {
+      return _i < _imax;
+    }
+    void next() {
+      _i++;
+    }
+    Node* current() {
+      return _node->fast_out(_i);;
+    }
+    UsesIteratorFast(DUIterator_Fast& imax, DUIterator_Fast& i, const Node* node)
+      : _imax(imax), _i(i), _node(node) {
+    }
+  };
+
+  class UsesIterator {
+    DUIterator& _i;
+    const Node* _node;
+
+  public:
+    bool cont() {
+      return _node->has_out(_i);
+    }
+    void next() {
+      _i++;
+    }
+    Node* current() {
+      return _node->out(_i);;
+    }
+    UsesIterator(DUIterator& i, const Node* node)
+      : _i(i), _node(node) {
+    }
+  };
+
+  // Iterate with i over all Proj uses calling callback
+  template<class Callback, class Iterator> ProjNode* apply_to_projs_any_iterator(Iterator i, Callback callback) const {
+    for (; i.cont(); i.next()) {
+      Node* p = i.current();
+      if (p->is_Proj()) {
+        ProjNode* proj = p->as_Proj();
+        if (callback(proj)) {
+          return proj;
+        }
+      } else {
+        assert(p == this && is_Start(), "else must be proj");
+      }
+    }
+    return nullptr;
+  }
```

Comment:
I was a little confused about this `apply_to_proj` construct. I this something we already use, a familiar concept, the `apply_to`?

Do we really iterate over `all` Proj uses? It seems that is not true if `callback` returns true, right?
This makes the semantic of the `apply_to_projs` a little subtle, right?

A suggestion: You could expect from `Callback` that it does not return `true/false`, but some `enum` value that describes what happens when you return that value. Maybe `ApplyToProjs::CONTINUE` vs `ApplyToProjs::BREAK_AND_RETURN_CURRENT_PROJ`?